sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBeanMapper.java

```diff
@@ -0,0 +1,250 @@
+            if (!(obj instanceof Document)) {
+                return false;
+            }
+            Document that = (Document) obj;
+            return this.id == that.id &&
+                    Objects.equals(this.name, that.name) &&
+                    Objects.equals(this.contents, that.contents);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(id, name, contents);
+        }
+    }
+
+    public static class Folder {
+        private int id;
+        private String name;
+        private List<Document> documents = new ArrayList<>();
+
+        public Folder() {}
+
+        public Folder(int id, String name, Document... documents) {
+            this.id = id;
+            this.name = name;
+            this.documents = Arrays.asList(documents);
+        }
+
+        public int getId() {
+            return id;
+        }
+
+        public void setId(int id) {
+            this.id = id;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        public List<Document> getDocuments() {
+            return documents;
+        }
+
+        public void setDocuments(List<Document> documents) {
+            this.documents = documents;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (!(obj instanceof Folder)) {
+                return false;
+            }
+            Folder that = (Folder) obj;
+            return this.id == that.id &&
+                    Objects.equals(this.name, that.name) &&
+                    Objects.equals(this.documents, that.documents);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(id, name, documents);
+        }
+    }
+
+    public interface DocumentDao extends GetHandle {
+        @SqlBatch("insert into folders (id, name) values (:id, :name)")
+        void insertFolders(@BindBean Folder... folders);
+
+        @SqlBatch("insert into documents (id, folder_id, name, contents) values (:d.id, :f.id, :d.name, :d.contents)")
+        void insertDocuments(@BindBean("f") Folder folder, @BindBean("d") Document... documents);
+
+        default Optional<Folder> getFolder(int folderId) {
+            return getHandle().createQuery(
+                    "select f.id f_id, f.name f_name, " +
+                            "d.id d_id, d.name d_name, d.contents d_contents " +
+                            "from folders f left join documents d " +
+                            "on f.id = d.folder_id " +
+                            "where f.id = :folderId")
+                    .bind("folderId", folderId)
+                    .registerRowMapper(BeanMapper.of(Folder.class, "f_"))
+                    .registerRowMapper(BeanMapper.of(Document.class, "d_"))
+                    .reduceRows(Optional.<Folder>empty(), (folder, row) -> {
```

Comment:
If `reduce` were still called `fold` this test case would be very hard to read :P